File: Computer-Science-50-Harward-University-edX/pset1/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    long long number;
    int digit, digit_prev, iter = 0, sum = 0;

    // getting card number

    number = get_long_long("Number: ");

    // adding digits acording to the rules
    while (number > 0)
    {
        digit_prev = digit;
        digit = number % 10;

        // every other digit by 2, starting with the number’s last digit
        if (iter % 2 == 0)
        {
            sum = sum + digit;
        }

        // every other digit by 2, starting with the number’s second-to-last digit
        if (iter % 2 == 1)
        {
            // first digit of the product
            sum = sum + 2 * digit % 10;
            // second digit of the product
            sum = sum + (2 * digit - 2 * digit % 10) / 10;
        }

        // shortening the number
        number = (number - digit) / 10;
        iter += 1;
    }

    // checking all conditions
    if (sum % 10 == 0 && digit == 3 && (digit_prev == 4 || digit_prev == 7) && iter == 15)
    {
        printf("AMEX\n");
    }

    else if (sum % 10 == 0 && digit == 5 && digit_prev >= 1 && digit_prev <= 5 && iter == 16)
    {
        printf("MASTERCARD\n");
    }

    else if (sum % 10 == 0 && digit == 4 && (iter == 13 || iter == 16))
    {
        printf("VISA\n");
    }

    else
    {
        printf("INVALID\n");
    }

}

File: Computer-Science-50-Harward-University-edX/pset6/credit.py
import cs50


def main():

    # infinite loop
    while True:

        # gets input from user
        number = cs50.get_int("Number: ")

        # if input incorrect - starts loop again
        if not number >= 0:
            continue

        number_str = str(number)
        length = len(number_str)

        # value for test
        test_sum = 0

        # secont-to-last digit
        digit_num = length - 2

        # executes algorithm for every second digit from secont-to-last
        while digit_num >= 0:
            # doubles every number in sequence
            product = str(int(number_str[digit_num]) * 2)
            # adds every digit of product to test_sum
            for i in product:
                test_sum += int(i)
            digit_num -= 2

        # last digit
        digit_num = length - 1

        # executes algorithm for every second digit form last
        while digit_num >= 0:
            # adds every digit to test_sum
            test_sum += int(number_str[digit_num])
            digit_num -= 2

        # True if test was successful
        test = test_sum % 10 == 0
        # two first digits of card number
        begin = int(number_str[0:2])

        # checking all conditions
        if test and (begin == 34 or begin == 37) and length == 15:
            print("AMEX")

        elif test and begin >= 51 and begin <= 5 and length == 16:
            print("MASTERCARD")

        elif test and str(begin)[0] == '4' and (length == 13 or length == 16):
            print("VISA")

        else:
            print("INVALID")

        exit(0)


# executes function
if __name__ == "__main__":
     main()